addons/stock/models/stock_move.py

```diff
@@ -1611,6 +1611,15 @@ def _action_assign(self, force_qty=False):
         moves_to_assign = self
         if not force_qty:
             moves_to_assign = self.filtered(lambda m: m.state in ['confirmed', 'waiting', 'partially_available'])
+
+        moves_mto = set()
+        for move in moves_to_assign:
+            if move.move_orig_ids:
+                moves_mto.add(move.id)
+
+        moves_mto = self.env['stock.move'].browse(moves_mto)
+        quants_cache = self.env['stock.quant']._get_quants_by_products_locations(moves_mto.product_id, moves_mto.location_id)
```

Comment:
No idea if this can have an impact, but we could add `[('quantity', '>', 0)]` as extra domain to reduce the number of quant to loop on

Reply:
It could be dangerous due to concurrency and duplicate quants (+x -y quants for the same key)